File: backend/app/schemas/tasks.py
from pydantic import BaseModel
from datetime import date as date_type, datetime


class TaskBase(BaseModel):
    title: str
    due_date: date_type
    priority: str = "medium"  # high, medium, low
    category: str = ""
    notes: str = ""
    done: bool = False
    assignees: list[str] = []


class TaskCreate(TaskBase):
    pass


class TaskUpdate(BaseModel):
    title: str | None = None
    due_date: date_type | None = None
    priority: str | None = None
    category: str | None = None
    notes: str | None = None
    done: bool | None = None
    assignees: list[str] | None = None


class TaskOut(TaskBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


class TaskStats:
    overdue: int
    today_count: int
    upcoming: int
    completed: int

File: frontend/src/api/tasks.ts
const API_BASE = import.meta.env.VITE_API_URL ?? "http://localhost:8000";
const ACCESS_KEY = "sparmanik_access_token";

function authHeaders(): HeadersInit {
  const token = localStorage.getItem(ACCESS_KEY);
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let msg = res.statusText;
    try {
      const d = await res.json();
      msg = d.detail ?? msg;
    } catch {}
    throw new Error(msg);
  }
  if (res.status === 204) return undefined as T;
  return res.json();
}

export type Priority = "high" | "medium" | "low";

export interface Task {
  id: number;
  title: string;
  due_date: string;
  priority: string;
  category: string;
  notes: string;
  done: boolean;
  assignees: string[];
  created_at: string;
}

export interface TaskCreate {
  title: string;
  due_date: string;
  priority: string;
  category: string;
  notes: string;
  done: boolean;
  assignees: string[];
}

export interface TaskUpdate {
  title?: string;
  due_date?: string;
  priority?: string;
  category?: string;
  notes?: string;
  done?: boolean;
  assignees?: string[];
}

export const tasksApi = {
  list() {
    return fetch(`${API_BASE}/api/tasks`, { headers: authHeaders() }).then((r) =>
      json<Task[]>(r)
    );
  },
  create(payload: TaskCreate) {
    return fetch(`${API_BASE}/api/tasks`, {
      method: "POST",
      headers: { ...authHeaders(), "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }).then((r) => json<Task>(r));
  },
  update(id: number, payload: TaskUpdate) {
    return fetch(`${API_BASE}/api/tasks/${id}`, {
      method: "PATCH",
      headers: { ...authHeaders(), "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }).then((r) => json<Task>(r));
  },
  remove(id: number) {
    return fetch(`${API_BASE}/api/tasks/${id}`, {
      method: "DELETE",
      headers: authHeaders(),
    }).then((r) => json<void>(r));
  },
};
